Add automatic image builder pipeline

DeinAlptraum opened the pull request:
This adds a Python script to extract via the Github API the most recent Hugo release that has not yet had a branch/image created, and a Github Actions CI that updates the Dockerfile to that version version, creates a branch and tag, and builds and pushes a new Docker image to the hub. 

The CI job runs once per day, cancelling itself if no new release has been found.

DeinAlptraum committed:
Add automatic image builder pipeline.

DeinAlptraum committed:
CI: Add Docker login and temporary testing repos

DeinAlptraum committed:
Prevent collision of tag and branch names

DeinAlptraum commented:
To see this in action, you may have a look at my fork: https://github.com/DeinAlptraum/docker-hugo/actions/workflows/main.yml
I've triggered the workflow manually for demonstrational purposes. Workflow #3 and #4 create images for the then-oldest non-published versions (after 0.88.1), which were 0.89.0 and 0.89.1 respectively. For #5 I've temporarily set the release list to empty, to show that this correctly cancels the job when no new release was found.

### Todos before this can be merged/used
1. Change my docker hub repo (nightoo/hugo) to yours (irgendwr/hugo/) 
2. Change the tags endpoint from my Github repo (DeinAlptraum/docker-hugo) to yours (irgendwr/docker-hugo)
3. Add the `DOCKER_USERNAME` and `DOCKERHUB_TOKEN` secrets on this repo
3.1 For this, create access token with Read & Write permission at https://app.docker.com/settings/personal-access-tokens/create

### Open Questions
1. How should this run? Imo the most reasonable thing would be to run this scheduled on a daily or weekly basis, as it is currently set.
5. How should we play catch-up here? There have been 166 releases since 0.88.1. One option would be to shorten the schedule to hourly or shorter for the start, then this should be up to date with v0.147.7 in about a week (or less). Or one could just keep triggering this manually. All options are a bit ugly, but should work. Note: I would have preferred for one job to build all "missing" releases sequentially, but GHA doesn't support loops and this would have been difficult to implement, just for this initial catch-up, so I've decided against this
6. Do you actually want an image for every version? We could also narrow this to only middle-version increases (0.88.0, 0.89.0, 0.90.0) etc., but note that this would mean we become unable to keep up with minor version increases: there is no way to know that one should have waited for 0.147.7 to be released before creating a new image for 0.147.0 at the time that was the latest.

irgendwr commented:
Thanks! I added the required secrets.

As for the questions:
1. Daily run sounds good to me. We can always tweak it if needed.
2. The catch-up doesn't have a high priority for me, I would even be fine with just having the new releases going forward. I think that we can decide on it after merging this.
3. Yes, it should definitely generate an image for every new version, including all patch releases.

irgendwr commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,46 @@
+        uses: actions/checkout@v4
+      - name: Determine latest un-imaged version
+        run: |
+          python get_latest_new_release.py
+          echo "version=$(cat _version.txt)" >> $GITHUB_ENV
```
I understand why you did it like this, but I'd prefer if the script didn't create a file and instead just printed the version to `stdout`, which you can then pipe to the env directly.

But since the info/debug prints are useful, they should stay, but end in `stderr` instead (which will be ignored by the pipe).

irgendwr commented on get_latest_new_release.py:
```diff
@@ -0,0 +1,59 @@
+print(releases_to_process)
+
+if releases_to_process:
+    minimum_new_version = min(releases_to_process, key=make_comparable)
```
I think it should be `max` instead of `min`. The latest version is the most important to imo, and is also what I would expect based on the name of the script.

irgendwr reviewed (changes requested):
Thank you for implementing this! :heart: 
I have some suggestions/requests but am looking very much forward to merging it :)

DeinAlptraum committed:
Address review comments

- build newest unbuilt release instead of oldest since v0.88.1
- write Python logging to stderr and use stdout for version output
- check available versions via Hugo repository's packages instead of releases

DeinAlptraum committed:
Fix GH_TOKEN passing

DeinAlptraum committed:
Fix default latest_Tag

DeinAlptraum committed:
Fix setting of GHA environment

DeinAlptraum committed:
Retrieve our tags from Docker Hub, stop using Github Tags

DeinAlptraum commented:
I've implemented your suggestions, including using Hugo's [released packages](https://github.com/gohugoio/hugo/pkgs/container/hugo), rather than the release page, to check for available versions (as discussed in private).

As a result, a few other changes also became necessary:
- due to `min` -> `max`, we cannot assume monotonicity anymore: we start with building the newest image, build our way downwards to the older versions, but then also have to jump back up every time a new version is released. I've thus adapted the script to get all our tags, rather than only the latest, and compare the full set with the tags on the Hugo repo.
- Due to this, we have to deal with the pagination stuff when collecting our own tags as well. I initially wanted to avoid this because I'm lazy, but there's no way around it now. But if we're having to deal with that anyway, we may as well get the images we made available directly from the Docker Hub API, and then not have to create tags on the Github repo just to keep an overview of what we've already processed.
- Github API needs a PAT for accessing packages... why... so the `GH_TOKEN` is now also passed to the Python script.
- At least the Github packages API returns _all_ available packages it seems, so no more pagination necessary on that side.
- Since determining which version to build next now depends on the API results from Docker Hub, rather than the tags on Github, I've switched the order and made the image upload the last step, so that this is blocked on repeating and failing the same version build in case the "Add & Commit" step fails

DeinAlptraum committed:
Regex filter our tags for version tags

Also format script

DeinAlptraum committed:
Use f-string for versioning script

DeinAlptraum committed:
Merge remote-tracking branch 'origin/master'

DeinAlptraum committed:
Update CI workflow for new Dockerfile

DeinAlptraum committed:
Fix regex matching

DeinAlptraum committed:
Update docker endpoints to irgendwr repo

irgendwr reviewed (approved):
:tada:
